src/ray/core_worker/core_worker.h

```diff
@@ -59,6 +59,99 @@ namespace core {
+  enum TaskStatusType { kPending, kRunning, kFinished };
+
+  /// This mutex should be used by caller to ensure consistency when transitioning
+  /// a task's state.
+  mutable absl::Mutex tasks_counter_mutex_;
+  absl::flat_hash_map<std::string, int> pending_tasks_counter_map_
+      GUARDED_BY(tasks_counter_mutex_);
+  absl::flat_hash_map<std::string, int> running_tasks_counter_map_
+      GUARDED_BY(tasks_counter_mutex_);
+  absl::flat_hash_map<std::string, int> finished_tasks_counter_map_
+      GUARDED_BY(tasks_counter_mutex_);
+
+  /// Tracks the total number of running tasks on this process.
+  int64_t running_total_ GUARDED_BY(tasks_counter_mutex_) = 0;
+
+  // Tracks tasks blocked in ray.get() and wait(). These counters overlap
+  // the counts of running_total_.
+  int64_t running_in_get_total_ GUARDED_BY(tasks_counter_mutex_) = 0;
+  int64_t running_in_wait_total_ GUARDED_BY(tasks_counter_mutex_) = 0;
+
+  void IncPending(const std::string &func_name) {
+    absl::MutexLock l(&tasks_counter_mutex_);
+    pending_tasks_counter_map_[func_name] += 1;
+  }
+
+  void MovePendingToRunning(const std::string &func_name) {
+    absl::MutexLock l(&tasks_counter_mutex_);
+    pending_tasks_counter_map_[func_name] -= 1;
+    running_tasks_counter_map_[func_name] += 1;
+    running_total_ += 1;
+    UpdateStats();
+  }
+
+  void MoveRunningToFinished(const std::string &func_name) {
+    absl::MutexLock l(&tasks_counter_mutex_);
+    running_tasks_counter_map_[func_name] -= 1;
+    finished_tasks_counter_map_[func_name] += 1;
+    running_total_ -= 1;
+    UpdateStats();
+  }
+
+  void SetMetricStatus(const std::string &func_name, rpc::TaskStatus status) {
+    absl::MutexLock l(&tasks_counter_mutex_);
+    // TODO(ekl) track and report per-name status.
+    if (status == rpc::TaskStatus::RUNNING_IN_RAY_GET) {
+      running_in_get_total_ += 1;
+    } else if (status == rpc::TaskStatus::RUNNING_IN_RAY_WAIT) {
+      running_in_wait_total_ += 1;
+    } else {
+      RAY_CHECK(false) << "Unexpected status " << rpc::TaskStatus_Name(status);
+    }
+    UpdateStats();
+  }
+
+  void UnsetMetricStatus(const std::string &func_name, rpc::TaskStatus status) {
+    absl::MutexLock l(&tasks_counter_mutex_);
+    if (status == rpc::TaskStatus::RUNNING_IN_RAY_GET) {
+      running_in_get_total_ -= 1;
+    } else if (status == rpc::TaskStatus::RUNNING_IN_RAY_WAIT) {
+      running_in_wait_total_ -= 1;
+    } else {
+      RAY_CHECK(false) << "Unexpected status " << rpc::TaskStatus_Name(status);
+    }
+    UpdateStats();
+  }
+
+  void UpdateStats() EXCLUSIVE_LOCKS_REQUIRED(&tasks_counter_mutex_) {
+    // Note that we set a Source=executor label so that metrics reported here don't
+    // conflict with metrics reported from task_manager.cc.
+    ray::stats::STATS_tasks.Record(
+        running_total_ - running_in_get_total_ - running_in_wait_total_,
+        {{"State", rpc::TaskStatus_Name(rpc::TaskStatus::RUNNING)},
+         {"Source", "executor"}});
```

Comment:
nit: I wonder if we should have some place/process to document all the labels on the metric? E.g. when it was declared.

Reply:
Added some comments in metric_defs.cc documenting the STATS_task definition and its labels.